File: StreamingBrowserElectron/src/sicherung.js
"use strict";

// Sicherung und Wiederherstellung.
//
// Der Grundsatz: hinein gehoert alles, was nicht wiederkommt, wenn es weg ist.
// Alles andere bleibt draussen.
//
// DRIN
//   settings     die Einstellungen
//   favorites    Watchlist, Weiterschauen, Mediathek, Verlauf, eigene Bilder
//   providers    die eigenen Anbieter
//   watchparty   welche Titel eingestellt und welche Runden beigetreten sind
//   sitzungen    die gemessenen Wiedergabesitzungen - der ganze Rueckblick
//   fassungen    welche Sprachfassung je Titel gemerkt ist
//   marken       Intro und Abspann je Staffel
//
// DRAUSSEN, und warum
//   Filterlisten, Geschmacksprofil und der Metadaten-Zwischenspeicher. Sie
//   sind zusammen ueber zehn Megabyte - mehr als alles andere zusammen - und
//   bauen sich von selbst wieder auf. Sie zu sichern hiesse, eine Sicherung
//   unbrauchbar gross zu machen fuer Daten, die niemand vermisst.
//
//   Der Spiegel des Geraeteabgleichs. Er beschreibt, was ein *anderes* Geraet
//   schon kennt; auf einem wiederhergestellten Stand waere er eine Behauptung
//   ueber eine Vergangenheit, die es dort nicht gab. Er baut sich beim ersten
//   Abgleich neu auf.
//
//   Die Geraetekennung. Sonst gaebe es nach dem Einlesen auf einem zweiten
//   Rechner zwei Geraete mit derselben Kennung, und das Relay haelt sie fuer
//   eines - Host-Wahl und Leiste waeren dahin.
//
// WAS EINMAL GEFEHLT HAT
//
// Bis Fassung 2 waren sitzungen, fassungen und marken nicht dabei. Das war
// still und teuer: die Sitzungen sind gemessene Zeit, die nie wiederkommt -
// beim Benutzer 224 Saetze ueber siebzehn Stunden -, und wer eine Sicherung
// einlas, verlor seinen ganzen Rueckblick, ohne dass irgendwo etwas davon
// stand. Fassungen und Marken sind Handarbeit je Titel und ebenso wenig
// nachzubauen.
//
// Das Modul rechnet nur; Dateien und Dialoge bleiben im Hauptprozess. Dadurch
// laesst sich das Heikle daran - was uebernommen wird und was nicht - ohne
// laufende App pruefen.

const KENNUNG = "elfix-sicherung";
// 2: sitzungen, fassungen und marken kamen dazu. Eine Sicherung der Fassung 1
// laesst sich weiter einlesen - was fehlt, bleibt dann eben stehen, statt
// geleert zu werden.
const FASSUNG = 2;

// Eine Sicherung aus dem laufenden Zustand.
//
// Die Geraetekennung bleibt bewusst draussen. Sonst gaebe es nach dem Einlesen
// auf einem zweiten Rechner zwei Geraete mit derselben Kennung, und das Relay
// haelt sie fuer eines - Host-Wahl und Leiste waeren dahin. Ein
// wiederhergestelltes Geraet erkennt das Relay ohnehin am Namen wieder und
// zieht seinen Stand auf die neue Kennung um.
function bauen({ settings, favorites, providers, watchparty, sitzungen, fassungen, marken,
  programm, anlass } = {}) {
  const einstellungen = settings ? JSON.parse(JSON.stringify(settings)) : null;
  if (einstellungen && einstellungen.watchparty) einstellungen.watchparty.deviceId = "";
  return {
    kennung: KENNUNG,
    fassung: FASSUNG,
    erstellt: new Date().toISOString(),
    programm: String(programm || ""),
    // Woher diese Sicherung kommt: von Hand, vor einem Update, vor dem
    // Einlesen einer anderen. Steht in der Datei, damit man ihr ansieht,
    // warum es sie gibt.
    anlass: String(anlass || "hand"),
    settings: einstellungen,
    favorites: Array.isArray(favorites) ? favorites : [],
    providers: Array.isArray(providers) ? providers : [],
    watchparty: watchparty || null,
    sitzungen: Array.isArray(sitzungen) ? sitzungen : [],
    // Fassungen und Marken sind Karten nach Titel, keine Listen.
    fassungen: fassungen && typeof fassungen === "object" ? fassungen : null,
    marken: marken && typeof marken === "object" ? marken : null
  };
}

// Ist das ueberhaupt eine Sicherung, und eine, die diese Fassung lesen kann?
function pruefen(daten) {
  if (!daten || typeof daten !== "object") return { ok: false, reason: "Das ist keine ELFIX-Sicherung" };
  if (daten.kennung !== KENNUNG) return { ok: false, reason: "Das ist keine ELFIX-Sicherung" };
  const fassung = Number(daten.fassung);
  if (!Number.isFinite(fassung) || fassung < 1) return { ok: false, reason: "Das ist keine ELFIX-Sicherung" };
  if (fassung > FASSUNG) {
    return { ok: false, reason: "Die Sicherung stammt aus einer neueren Version von ELFIX" };
  }
  return { ok: true };
}

// Wie viel steckt drin? Das steht im Dialog vor dem Einlesen - damit niemand
// eine fast leere Sicherung ueber eine volle Ablage legt, ohne es vorher zu
// sehen.
function umfang(daten) {
  const eintraege = Array.isArray(daten?.favorites) ? daten.favorites : [];
  const zahl = (wert) => (Number.isFinite(Number(wert)) && Number(wert) > 0 ? Number(wert) : 0);
  const zaehle = (wert) => (wert && typeof wert === "object" ? Object.keys(wert).length : 0);
  return {
    favoriten: eintraege.length,
    bilder: eintraege.filter((eintrag) => eintrag && eintrag.customThumbnail).length,
    weiterschauen: eintraege.filter((eintrag) => zahl(eintrag?.position) > 0 || zahl(eintrag?.currentTime) > 0).length,
    anbieter: Array.isArray(daten?.providers) ? daten.providers.length : 0,
    raeume: Array.isArray(daten?.settings?.watchparty?.rooms) ? daten.settings.watchparty.rooms.length : 0,
    einstellungen: Boolean(daten?.settings),
    // Eine Sicherung der Fassung 1 kennt diese drei nicht. Dann steht dort
    // null - und der Dialog kann sagen, dass sie nichts davon enthaelt,
    // statt eine Null zu behaupten.
    sitzungen: Array.isArray(daten?.sitzungen) ? daten.sitzungen.length : null,
    fassungen: daten && "fassungen" in daten ? zaehle(daten.fassungen) : null,
    marken: daten && "marken" in daten ? zaehle(daten.marken) : null
  };
}

// Ein Satz darueber, was fehlt.
//
// Eine Sicherung der Fassung 1 hat keine Sitzungen, keine Fassungen und keine
// Marken - nicht, weil sie leer waeren, sondern weil es sie dort nicht gab.
// Der Unterschied gehoert vor das Einlesen: was hier steht, bleibt stehen,
// statt geleert zu werden.
function fehlendeTeile(daten) {
  const fehlt = [];
  if (!daten || !("sitzungen" in daten)) fehlt.push("Wiedergabezeiten");
  if (!daten || !("fassungen" in daten)) fehlt.push("gemerkte Sprachfassungen");
  if (!daten || !("marken" in daten)) fehlt.push("Intromarken");
  return fehlt;
}

// Die Einstellungen, wie sie beim Einlesen gelten sollen: alles aus der
// Sicherung, aber mit der Kennung dieses Rechners. Fehlt hier noch eine, bleibt
// sie leer - dann vergibt die App beim naechsten Verbinden eine neue.
function einstellungenUebernehmen(ausSicherung, eigeneKennung) {
  if (!ausSicherung) return null;
  const uebernommen = JSON.parse(JSON.stringify(ausSicherung));
  uebernommen.watchparty = { ...(uebernommen.watchparty || {}), deviceId: String(eigeneKennung || "") };
  return uebernommen;
}

// Ein Dateiname mit Datum - beim zweiten Mal am selben Tag fragt der Dialog von
// selbst nach dem Ueberschreiben.
function dateiname(jetzt = new Date()) {
  const zwei = (wert) => String(wert).padStart(2, "0");
  return `ELFIX-Sicherung-${jetzt.getFullYear()}-${zwei(jetzt.getMonth() + 1)}-${zwei(jetzt.getDate())}.elfix.json`;
}

// Der Name einer Sicherung, die die App selbst anlegt.
//
// Mit Uhrzeit, nicht nur mit Datum: vor einem Update kann zweimal am selben
// Tag eine entstehen, und die zweite darf die erste nicht ueberschreiben -
// sonst waere die Rueckfahrkarte weg, sobald man sie zweimal braucht.
function selbstName(anlass, jetzt = new Date()) {
  const zwei = (wert) => String(wert).padStart(2, "0");
  const stempel = `${jetzt.getFullYear()}${zwei(jetzt.getMonth() + 1)}${zwei(jetzt.getDate())}`
    + `-${zwei(jetzt.getHours())}${zwei(jetzt.getMinutes())}${zwei(jetzt.getSeconds())}`;
  const sauber = String(anlass || "auto").replace(/[^a-z0-9-]/gi, "").toLowerCase() || "auto";
  return `ELFIX-${sauber}-${stempel}.elfix.json`;
}

// Welche selbst angelegten Sicherungen weg duerfen.
//
// Sie sammeln sich sonst: vor jedem Update eine, und ELFIX bekommt mehrere
// Fassungen in der Woche. Behalten wird die juengste Handvoll - genug, um ein
// misslungenes Update zu ueberstehen, und wenig genug, dass der Datenordner
// nicht zulaeuft.
//
// Sortiert wird nach dem Namen, und das genuegt: der Zeitstempel darin ist so
// gebaut, dass alphabetisch und chronologisch dasselbe ist.
function altePutzen(namen, behalten = 5) {
  const eigene = (Array.isArray(namen) ? namen : [])
    .filter((name) => /^ELFIX-[a-z0-9-]+-\d{8}-\d{6}\.elfix\.json$/i.test(String(name)))
    .sort();
  const wieViele = Math.max(0, Number(behalten) || 0);
  return eigene.slice(0, Math.max(0, eigene.length - wieViele));
}

module.exports = {
  KENNUNG, FASSUNG, bauen, pruefen, umfang, fehlendeTeile,
  einstellungenUebernehmen, dateiname, selbstName, altePutzen
};

File: android/app/src/main/java/local/elflix/android/Sicherung.java
package local.elflix.android;

import android.content.Context;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * Sicherungen auf Android.
 *
 * <h2>Wozu</h2>
 *
 * <p>Ein Update laesst den Bestand in aller Regel stehen - die APK wird
 * darueber installiert, nicht neu. In aller Regel ist aber nicht immer, und
 * der eine Fall, in dem es schiefgeht, ist genau der, in dem niemand eine
 * Sicherung hat. Deshalb legt ELFIX vor jeder Installation selbst eine an.
 *
 * <p>Sie kostet einen Augenblick und ein paar hundert Kilobyte. Sie zu haben
 * und nicht zu brauchen ist der bessere Handel.
 *
 * <h2>Was hineingehoert</h2>
 *
 * <p>Entschieden wird das in {@code sicherung.js} - demselben Modul, mit dem
 * der Rechner seine Sicherungen baut und liest. Eine hier gebaute Sicherung
 * laesst sich damit am Rechner einlesen und umgekehrt; zwei Formate waeren
 * zwei Wahrheiten.
 *
 * <h2>Wo sie liegt</h2>
 *
 * <p>Im eigenen Datenordner der App, Unterordner {@code sicherungen}. Nicht im
 * gemeinsamen Speicher: dort braeuchte es eine Berechtigung, und eine
 * Berechtigungsfrage vor einem Update ist genau die Huerde, an der eine
 * Sicherheitskopie scheitert. Ein Update ueberlebt diesen Ordner - eine
 * Deinstallation nicht, und das ist der Preis.
 */
public final class Sicherung {
    private static final String TAG = CrashReporter.TAG;
    private static final String ORDNER = "sicherungen";

    private final Context context;
    private final Kern kern;
    private final Bestand bestand;
    private final Statistik statistik;
    private final Watchparty watchparty;

    public Sicherung(Context context, Kern kern, Bestand bestand, Statistik statistik,
                     Watchparty watchparty) {
        this.context = context.getApplicationContext();
        this.kern = kern;
        this.bestand = bestand;
        this.statistik = statistik;
        this.watchparty = watchparty;
    }

    /** Die Fassung dieser App - fuer die Datei, damit man ihr ansieht, woher sie kommt. */
    private String programmFassung() {
        try {
            return context.getPackageManager()
                .getPackageInfo(context.getPackageName(), 0).versionName;
        } catch (Exception fehler) {
            return "";
        }
    }

    /** Wird gerufen, wenn die Sicherung steht - oder feststeht, dass keine kommt. */
    public interface Fertig {
        void sicherungFertig(String pfad);
    }

    /**
     * Eine Sicherung anlegen.
     *
     * <p>Ohne Nachfrage und ohne Abbruch: schlaegt sie fehl, geht es trotzdem
     * weiter. Eine Sicherung soll ein Update begleiten, nicht verhindern.
     *
     * @param anlass warum es sie gibt - steht im Dateinamen und in der Datei
     */
    public void anlegen(String anlass, Fertig fertig) {
        if (kern == null || !kern.istBereit()) {
            Log.d(TAG, "Sicherung: der Kern steht nicht - keine angelegt");
            if (fertig != null) fertig.sicherungFertig("");
            return;
        }
        JSONObject quellen = new JSONObject();
        try {
            quellen.put("settings", einstellungen());
            quellen.put("favorites", bestand == null ? new JSONArray() : bestand.roh());
            quellen.put("providers", anbieterSichern(ProviderStore.ladeAlle(context)));
            quellen.put("watchparty", watchparty == null ? null : watchparty.kontoSatz());
            // Die Sitzungen sind gemessene Zeit und kommen nie wieder.
            quellen.put("sitzungen", statistik == null ? new JSONArray() : statistik.alle());
            // Fassungen und Marken liegen in ihren eigenen Ablagen. Von dort
            // gelesen und nicht ueber ihre Klassen: die halten ihren Stand im
            // Kern, und dort ihn abzufragen waere ein zweiter Weg zu denselben
            // Daten - mit der Gefahr, dass er ein anderes Ergebnis liefert.
            quellen.put("fassungen", new Ablage(context, "fassungen.json", 1).lesen());
            quellen.put("marken", new Ablage(context, "marken.json", 1).lesen());
            quellen.put("programm", programmFassung());
            quellen.put("anlass", anlass == null ? "auto" : anlass);
        } catch (Exception fehler) {
            Log.e(TAG, "Sicherung liess sich nicht vorbereiten", fehler);
            if (fertig != null) fertig.sicherungFertig("");
            return;
        }
        kern.rufe("sicherung.bauen", Kern.args(quellen), (wert, fehler) -> {
            if (fehler != null || wert == null) {
                Log.e(TAG, "Sicherung nicht gebaut: " + fehler);
                if (fertig != null) fertig.sicherungFertig("");
                return;
            }
            String pfad = schreiben(anlass, wert);
            if (fertig != null) fertig.sicherungFertig(pfad);
        });
    }

    /**
     * Die Einstellungen dieses Geraets, so wie sie in eine Sicherung gehoeren.
     *
     * <p>Die Watchparty gehoert dazu - Raeume und Adresse -, die Geraetekennung
     * ausdruecklich nicht. Sie gehoert dem Geraet: zwei Geraete mit derselben
     * Kennung gelten im Raum als eines.
     */
    private JSONObject einstellungen() {
        JSONObject alles = new JSONObject();
        try {
            JSONObject wp = new JSONObject();
            if (watchparty != null) {
                wp.put("enabled", watchparty.istEingeschaltet());
                wp.put("serverUrl", watchparty.serverUrl());
                wp.put("deviceName", watchparty.geraetName());
                JSONArray codes = new JSONArray();
                for (String code : watchparty.raumcodes()) codes.put(code);
                wp.put("rooms", codes);
            }
            alles.put("watchparty", wp);
            // Die uebrigen Einstellungen der App - Startseite, Werbeblocker,
            // Autoplay, Geraeteabgleich. Sie liegen in SharedPreferences und
            // gehen als Ganzes mit; welcher Schluessel was bedeutet, weiss die
            // App beim Einlesen selbst.
            JSONObject android = new JSONObject();
            for (java.util.Map.Entry<String, ?> eintrag
                : context.getSharedPreferences("elflix_settings", Context.MODE_PRIVATE)
                    .getAll().entrySet()) {
                android.put(eintrag.getKey(), eintrag.getValue());
            }
            alles.put("android", android);
        } catch (Exception fehler) {
            Log.e(TAG, "Einstellungen nicht gesammelt", fehler);
        }
        return alles;
    }

    /**
     * Anbieter in derselben Form sichern, in der sie auch abgelegt werden.
     *
     * <p>{@link Provider#alsJson()} reicht hier nicht: sie ist die reduzierte
     * Form fuer den gemeinsamen Kern und enthaelt weder den Schalter noch den
     * Werbefilter-Stand. Eine Sicherung muss beide beim Einlesen erhalten.
     */
    static JSONArray anbieterSichern(List<Provider> anbieter) {
        JSONArray liste = new JSONArray();
        if (anbieter == null) return liste;
        for (Provider eintrag : anbieter) {
            if (eintrag != null) liste.put(eintrag.alsAblage());
        }
        return liste;
    }

    /** Die fertige Sicherung ablegen und alte wegraeumen. */
    private String schreiben(String anlass, String inhalt) {
        try {
            File ordner = new File(context.getFilesDir(), ORDNER);
            if (!ordner.isDirectory() && !ordner.mkdirs()) {
                Log.w(TAG, "Sicherungsordner nicht angelegt");
                return "";
            }
            String text = inhalt;
            // Der Kern gibt JSON-Text zurueck; er ist schon fertig.
            File ziel = new File(ordner, namen(anlass));
            java.io.FileOutputStream aus = new java.io.FileOutputStream(ziel);
            try {
                aus.write(text.getBytes("UTF-8"));
                aus.flush();
            } finally {
                aus.close();
            }
            Log.i(TAG, "Sicherung angelegt: " + ziel.getName() + " (" + ziel.length() + " Bytes)");
            aufraeumen(ordner);
            return ziel.getAbsolutePath();
        } catch (Exception fehler) {
            // Kein Abbruch. Siehe oben.
            Log.e(TAG, "Sicherung nicht geschrieben", fehler);
            return "";
        }
    }

    /**
     * Der Name - mit Uhrzeit, nicht nur mit Datum.
     *
     * <p>Vor einem Update kann zweimal am selben Tag eine entstehen, und die
     * zweite darf die erste nicht ueberschreiben: sonst waere die
     * Rueckfahrkarte weg, sobald man sie zweimal braucht. Dieselbe Form wie am
     * Rechner ({@code selbstName} in sicherung.js).
     */
    private static String namen(String anlass) {
        java.util.Calendar jetzt = java.util.Calendar.getInstance();
        String stempel = String.format(java.util.Locale.ROOT, "%04d%02d%02d-%02d%02d%02d",
            jetzt.get(java.util.Calendar.YEAR), jetzt.get(java.util.Calendar.MONTH) + 1,
            jetzt.get(java.util.Calendar.DAY_OF_MONTH), jetzt.get(java.util.Calendar.HOUR_OF_DAY),
            jetzt.get(java.util.Calendar.MINUTE), jetzt.get(java.util.Calendar.SECOND));
        String sauber = (anlass == null ? "auto" : anlass).replaceAll("[^a-zA-Z0-9-]", "").toLowerCase();
        if (sauber.isEmpty()) sauber = "auto";
        return "ELFIX-" + sauber + "-" + stempel + ".elfix.json";
    }

    /**
     * Alte Sicherungen wegraeumen.
     *
     * <p>Sie sammeln sich sonst: vor jedem Update eine, und ELFIX bekommt
     * mehrere Fassungen in der Woche. Behalten wird die juengste Handvoll -
     * genug fuer ein misslungenes Update, wenig genug, dass der Datenordner
     * nicht zulaeuft. Sortiert wird nach dem Zeitstempel, weil der Anlass vor
     * ihm im Namen steht.
     */
    private static void aufraeumen(File ordner) {
        String[] namen = ordner.list();
        if (namen == null || namen.length <= BEHALTEN) return;
        List<String> eigene = new ArrayList<>();
        for (String name : namen) {
            if (name != null && name.startsWith("ELFIX-") && name.endsWith(".elfix.json")) {
                eigene.add(name);
            }
        }
        if (eigene.size() <= BEHALTEN) return;
        for (String name : alteSicherungen(eigene, BEHALTEN)) {
            File weg = new File(ordner, name);
            if (!weg.delete()) Log.d(TAG, "Alte Sicherung blieb liegen: " + name);
        }
    }

    /**
     * Die Namen tragen vor dem Zeitstempel einen Anlass. Deshalb ist die volle
     * alphabetische Reihenfolge nicht chronologisch, sobald Sicherungen aus
     * verschiedenen Anlaessen zusammenkommen.
     */
    static List<String> alteSicherungen(List<String> namen, int behalten) {
        List<String> sortiert = new ArrayList<>(namen == null ? java.util.Collections.emptyList() : namen);
        sortiert.sort(nachZeitstempel());
        int anzahl = Math.max(0, behalten);
        return new ArrayList<>(sortiert.subList(0, Math.max(0, sortiert.size() - anzahl)));
    }

    private static String zeitstempel(String name) {
        java.util.regex.Matcher treffer = java.util.regex.Pattern
            .compile("-([0-9]{8}-[0-9]{6})\\.elfix\\.json$")
            .matcher(name == null ? "" : name);
        return treffer.find() ? treffer.group(1) : "";
    }

    private static Comparator<String> nachZeitstempel() {
        return (links, rechts) -> {
            int vergleich = zeitstempel(links).compareTo(zeitstempel(rechts));
            return vergleich != 0 ? vergleich : links.compareTo(rechts);
        };
    }

    /** Wie viele selbst angelegte Sicherungen stehenbleiben. */
    private static final int BEHALTEN = 5;

    /** Die vorhandenen Sicherungen - fuer die Anzeige in den Einstellungen. */
    public List<File> vorhandene() {
        List<File> liste = new ArrayList<>();
        File ordner = new File(context.getFilesDir(), ORDNER);
        File[] dateien = ordner.listFiles();
        if (dateien == null) return liste;
        java.util.Arrays.sort(dateien, (links, rechts) ->
            nachZeitstempel().compare(rechts.getName(), links.getName()));
        for (File datei : dateien) {
            if (datei.isFile() && datei.getName().endsWith(".elfix.json")) liste.add(datei);
        }
        return liste;
    }
}
